
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://seqscope.github.io/NovaScope/getting_started/job_config/">
      
      
        <link rel="prev" href="../access_data/">
      
      
        <link rel="next" href="../execute/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.5.3, mkdocs-material-9.5.13">
    
    
      
        <title>Job Configuration - NovaScope</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.7e359304.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../css/custom.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#configuring-a-novascope-run" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="NovaScope" class="md-header__button md-logo" aria-label="NovaScope" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            NovaScope
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Job Configuration
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1M8 13h8v-2H8v2m9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1 0 1.71-1.39 3.1-3.1 3.1h-4V17h4a5 5 0 0 0 5-5 5 5 0 0 0-5-5Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5M7 15a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/seqscope/NovaScope" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    seqscope/NovaScope
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../.." class="md-tabs__link">
          
  
    
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../installation/requirement/" class="md-tabs__link">
          
  
    
  
  Installation

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../intro/" class="md-tabs__link">
          
  
    
  
  Getting Started

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../walkthrough/intro/" class="md-tabs__link">
          
  
    
  
  NovaScope Walkthrough

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="NovaScope" class="md-nav__button md-logo" aria-label="NovaScope" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54Z"/></svg>

    </a>
    NovaScope
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/seqscope/NovaScope" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.1 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2023 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    seqscope/NovaScope
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../.." class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Home
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_1" id="__nav_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../home/documentation_overview/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Documentation Overview
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../home/workflow_structure/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Workflow Structure
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Installation
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Installation
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../installation/requirement/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installing NovaScope
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../installation/env_setup/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Environment Setup
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../installation/slurm/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Slurm
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../installation/requirement_for_plus/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Additional Installations
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Getting Started
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Getting Started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../intro/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../access_data/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Accessing Input Data
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Job Configuration
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Job Configuration
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#preparing-input-config-files" class="md-nav__link">
    <span class="md-ellipsis">
      Preparing Input Config Files
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Preparing Input Config Files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#a-template-of-the-config-file" class="md-nav__link">
    <span class="md-ellipsis">
      A Template of the Config File
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#detailed-description-of-individual-fields" class="md-nav__link">
    <span class="md-ellipsis">
      Detailed Description of Individual Fields
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Detailed Description of Individual Fields">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input" class="md-nav__link">
    <span class="md-ellipsis">
      Input
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#requests" class="md-nav__link">
    <span class="md-ellipsis">
      Requests
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#upstream-downstream" class="md-nav__link">
    <span class="md-ellipsis">
      Upstream &amp; Downstream
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../execute/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Executing NovaScope
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../output/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Output
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    NovaScope Walkthrough
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            NovaScope Walkthrough
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../walkthrough/intro/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Introduction
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Rule Execution Guide
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            Rule Execution Guide
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../walkthrough/execution_guide/core_concepts/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Core Concepts
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../walkthrough/execution_guide/rule_execution/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Execution Flow by Request
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3" >
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Main Rules
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            Main Rules
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../walkthrough/rules/fastq2sbcd/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    fastq2sbcd
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../walkthrough/rules/sbcd2chip/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    sbcd2chip
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../walkthrough/rules/smatch/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    smatch
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../walkthrough/rules/align/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    align
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../walkthrough/rules/dge2sdge/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    dge2sdge
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../walkthrough/rules/sdge_visual/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    sdge_visual
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_4" >
        
          
          <label class="md-nav__link" for="__nav_4_4" id="__nav_4_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Plus Rules
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_4">
            <span class="md-nav__icon md-icon"></span>
            Plus Rules
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../walkthrough/rules/historef/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    historef
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../walkthrough/rules/sdge2sdgeAR/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    sdge2sdgeAR
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../walkthrough/rules/sdgeAR_reformat/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    sdgeAR_reformat
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../walkthrough/rules/sdgeAR_segment/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    sdgeAR_segment
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#preparing-input-config-files" class="md-nav__link">
    <span class="md-ellipsis">
      Preparing Input Config Files
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Preparing Input Config Files">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#a-template-of-the-config-file" class="md-nav__link">
    <span class="md-ellipsis">
      A Template of the Config File
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#detailed-description-of-individual-fields" class="md-nav__link">
    <span class="md-ellipsis">
      Detailed Description of Individual Fields
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Detailed Description of Individual Fields">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#input" class="md-nav__link">
    <span class="md-ellipsis">
      Input
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#requests" class="md-nav__link">
    <span class="md-ellipsis">
      Requests
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#upstream-downstream" class="md-nav__link">
    <span class="md-ellipsis">
      Upstream &amp; Downstream
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  

  
  


<h1 id="configuring-a-novascope-run">Configuring a NovaScope Run<a class="headerlink" href="#configuring-a-novascope-run" title="Permanent link">&para;</a></h1>
<h2 id="overview">Overview<a class="headerlink" href="#overview" title="Permanent link">&para;</a></h2>
<p>Once you have <a href="../../installation/requirement/">installed NovaScope</a> and <a href="../access_data/">downloaded the input data</a>, the next step is to configure a <a href="https://seqscope.github.io/NovaScope/">NovaScope</a> run. This mainly involves preparing the input job configuration file (in YAML, <code>config_job.yaml</code>) for the run.</p>
<h2 id="preparing-input-config-files">Preparing Input Config Files<a class="headerlink" href="#preparing-input-config-files" title="Permanent link">&para;</a></h2>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>The pipeline requires to have <code>config_job.yaml</code> file in the working directory, which will be indicated by <code>-d</code> or <code>--directory</code> when executing NovaScope, to specify all input files, output files, and parameters. </p>
</div>
<p>For user's convenience, we provide separate example <code>config_job.yaml</code> files for the <a href="https://github.com/seqscope/NovaScope/blob/main/testrun/minmal_test_run/config_job.yaml">Minimal Test Run Dataset</a>, <a href="https://github.com/seqscope/NovaScope/blob/main/testrun/shallow_liver_section/config_job.yaml">Shallow Liver Section Dataset</a>, and <a href="https://github.com/seqscope/NovaScope/blob/main/testrun/deep_liver_section/config_job.yaml">Deep Liver Section Dataset</a> test runs.</p>
<p>The details of each item specified in the <code>config_job.yaml</code> is described below:</p>
<h3 id="a-template-of-the-config-file">A Template of the Config File<a class="headerlink" href="#a-template-of-the-config-file" title="Permanent link">&para;</a></h3>
<p>Below is a template of the <code>config_job.yaml</code> file. </p>
<p>Mandatory fields are marked as "REQUIRED FIELD".</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">  1</span>
<span class="normal">  2</span>
<span class="normal">  3</span>
<span class="normal">  4</span>
<span class="normal">  5</span>
<span class="normal">  6</span>
<span class="normal">  7</span>
<span class="normal">  8</span>
<span class="normal">  9</span>
<span class="normal"> 10</span>
<span class="normal"> 11</span>
<span class="normal"> 12</span>
<span class="normal"> 13</span>
<span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1">## Section to Specify Input Datta</span>
<span class="nt">input</span><span class="p">:</span>
<span class="w">  </span><span class="nt">flowcell</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;flowcell_id&gt;</span><span class="w">                       </span><span class="c1">## REQUIRED FIELD (e.g. N3-HG5MC)</span>
<span class="w">  </span><span class="nt">chip</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;chip_id&gt;</span><span class="w">                               </span><span class="c1">## REQUIRED FIELD (e.g. B08C)</span>
<span class="w">  </span><span class="nt">species</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;species_info&gt;</span><span class="w">                       </span><span class="c1">## REQUIRED FIELD (e.g. &quot;mouse&quot;)</span>
<span class="w">  </span><span class="nt">lane</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;lane_id&gt;</span><span class="w">                               </span><span class="c1">## Optional. Defaults to auto-assignment from the last character of chip_id (A-&gt;1, B-&gt;2, C-&gt;3, D-&gt;4) if absent.</span>
<span class="w">  </span><span class="nt">seq1st</span><span class="p">:</span><span class="w">                                       </span><span class="c1">## 1st-seq information</span>
<span class="w">    </span><span class="nt">id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;seq1st_id&gt;</span><span class="w">                             </span><span class="c1">## Optional. Defaults to &quot;L{lane}&quot; if absent.</span>
<span class="w">    </span><span class="nt">fastq</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;path_to_seq1st_fastq_file&gt;</span><span class="w">          </span><span class="c1">## REQUIRED FIELD</span>
<span class="w">    </span><span class="nt">layout</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;path_to_sbcd_layout&gt;</span><span class="w">               </span><span class="c1">## Optional. Default based on chip_id</span>
<span class="w">  </span><span class="nt">seq2nd</span><span class="p">:</span><span class="w">                                       </span><span class="c1">## 2nd-seq information. See the &quot;Detailed Description of Individual Fields&quot; below.</span>
<span class="w">    </span><span class="c1">## specify the first pair of FASTQs.</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;seq2nd_pair1_id&gt;</span><span class="w">                     </span><span class="c1">## Optional. If provided, must be unique among all 2nd-seq pairs. Defaults to automatic assignment based on fastq_R1 if absent (see details below).</span>
<span class="w">      </span><span class="nt">fastq_R1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;path_to_seq2nd_pair1_fastq_Read1_file&gt;</span><span class="w"> </span><span class="c1">## REQUIRED FIELD - path to Read 1 FASTQ file,</span>
<span class="w">      </span><span class="nt">fastq_R2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;path_to_seq2nd_pair1_fastq_Read2_file&gt;</span><span class="w"> </span><span class="c1">## REQUIRED FIELD - path to Read 2 FASTQ file</span>
<span class="w">    </span><span class="c1">## if there is a second pair of FASTQs ...</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;seq2nd_pair2_id&gt;</span><span class="w">                     </span>
<span class="w">      </span><span class="nt">fastq_R1</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;path_to_seq2nd_pair2_fastq_Read1_file&gt;</span>
<span class="w">      </span><span class="nt">fastq_R2</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;path_to_seq2nd_pair2_fastq_Read2_file&gt;</span>
<span class="w">    </span><span class="c1">## ... (if there are more 2nd-seq FASTQ files)</span>
<span class="w">  </span><span class="nt">run_id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;run_id&gt;</span><span class="w">                              </span><span class="c1">## Optional. See the &quot;Detailed Description of Individual Fields&quot; below.</span>
<span class="w">  </span><span class="nt">unit_id</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;unit_id&gt;</span><span class="w">                            </span><span class="c1">## Optional. See the &quot;Detailed Description of Individual Fields&quot; below.</span>
<span class="w">  </span><span class="nt">histology</span><span class="p">:</span><span class="w">                                    </span><span class="c1">## Histology information. Only required if histology alignment is needed. See the &quot;Detailed Description of Individual Fields&quot; below.</span>
<span class="w">    </span><span class="c1">## specify the first input histology file</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;path_to_1st_histology_file&gt;</span><span class="w">        </span><span class="c1">## REQUIRED FIELD - path to the input histology file</span>
<span class="w">      </span><span class="nt">magnification</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;magnification&gt;</span><span class="w">            </span><span class="c1">## Optional - specify the magnification of the input histology file, default is &quot;10X&quot;</span>
<span class="w">      </span><span class="nt">figtype</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;type&gt;</span><span class="w">                           </span><span class="c1">## Optional - specify the type of the histology file. Options: &quot;hne&quot;, &quot;dapi&quot;, and &quot;fl&quot;. </span>
<span class="w">    </span><span class="c1">## if there is a second input histology file ...</span>
<span class="w">    </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="nt">path</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;path_to_2nd_histology_file&gt;</span><span class="w">       </span>
<span class="w">      </span><span class="nt">magnification</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;magnification&gt;</span><span class="w">                         </span>
<span class="w">      </span><span class="nt">figtype</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;type&gt;</span><span class="w">   </span>
<span class="w">    </span><span class="c1">## ...                           </span>
<span class="c1">## Output</span>
<span class="nt">output</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;output_directory&gt;</span><span class="w">                      </span><span class="c1">## REQUIRED FIELD (e.g. /path/to/output/directory)</span>
<span class="nt">request</span><span class="p">:</span><span class="w">                                        </span><span class="c1">## See the &quot;Detailed Description of Individual Fields&quot; below.</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;required_output1&gt;</span><span class="w">                          </span><span class="c1">## REQUIRED FIELD (e.g. sge-per-run)</span>
<span class="w">  </span><span class="p p-Indicator">-</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;required_output2&gt;</span><span class="w">                          </span><span class="c1">## Optionally, you can request multiple outputs</span>
<span class="w">  </span><span class="c1"># ...</span>

<span class="c1">## Environment</span>
<span class="nt">env_yml</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">&lt;path_to_config_env.yaml_file&gt;</span><span class="w">         </span><span class="c1">## If absent, NovaScope use the &quot;config_env.yaml&quot; file in the `info` subdirectory in the Novascope repository.</span>

<span class="c1">## ================================================</span>
<span class="c1">##</span>
<span class="c1">##  Additional Fields:</span>
<span class="c1">## </span>
<span class="c1">##    The &quot;upstream&quot;, &quot;histology&quot;, and &quot;downstream&quot; parameters are included below, along side the default values.</span>
<span class="c1">##    Revise and enable the following parameters ONLY IF you wish to utilize values different than the default.</span>
<span class="c1">##</span>
<span class="c1">## ================================================</span>

<span class="c1">### UNCOMMENT RELEVANT LINES TO ENABLE THE ADDITIONAL PARAMETERS</span>
<span class="c1">#upstream:</span>
<span class="c1">#  fastq2sbcd:</span>
<span class="c1">#    format: DraI32          ## Example data uses DraI31, but DraI32 is a typical format.</span>
<span class="c1">#</span>
<span class="c1">#  sbcd2chip:                ## specify the parameters for sbcd2chip</span>
<span class="c1">#    gap_row: 0.0517</span>
<span class="c1">#    gap_col: 0.0048</span>
<span class="c1">#    dup_maxnum: 1</span>
<span class="c1">#    dup_maxdist: 1</span>
<span class="c1">#</span>
<span class="c1">#  smatch:                   ## specify the parameters for smatch</span>
<span class="c1">#    skip_sbcd: 0            ## If absent, default skip_sbcd follows the fastq2sbcd format: 1 for DraI31 and 0 for DraI32.</span>
<span class="c1">#    match_len: 27           ## Length of spatial barcode considered to be a perfect match.</span>
<span class="c1">#</span>
<span class="c1">#  align:                    ## specify the parameters for align (STARsolo)</span>
<span class="c1">#    min_match_len: 30       ## A minimum number of matching bases.</span>
<span class="c1">#    min_match_frac: 0.66    ## A minimum fraction of matching bases.</span>
<span class="c1">#    len_sbcd: 30            ## Length of spatial barcode (in Read 1) to be copied to output FASTQ file (Read 1).</span>
<span class="c1">#    len_umi: 9              ## Length of UMI barcode (in Read 2) to be copied to output FASTQ file (Read 1).</span>
<span class="c1">#    len_r2: 101             ## Length of read 2 after trimming (including randomers).</span>
<span class="c1">#    exist_action: overwrite ## Skip the action or overwrite the file if an intermediate or output file already exists. Options: &quot;skip&quot;, and &quot;overwrite&quot;.</span>
<span class="c1">#    resource:               ## See the &quot;Detailed Description of Individual Fields&quot; below.</span>
<span class="c1">#      assign_type: stdin</span>
<span class="c1">#      stdin:</span>
<span class="c1">#        partition: standard</span>
<span class="c1">#        threads: 10</span>
<span class="c1">#        memory: 70000m</span>
<span class="c1">#</span>
<span class="c1">#  visualization:            ## specify the parameters for visualization</span>
<span class="c1">#    drawxy:                 ## specify the parameters for visualization for sbcd and smatch images</span>
<span class="c1">#      coord_per_pixel: 1000</span>
<span class="c1">#      intensity_per_obs: 50</span>
<span class="c1">#      icol_x: 3</span>
<span class="c1">#      icol_y: 4</span>
<span class="c1">#    drawsge:                ## specify the parameters for sdge visualization </span>
<span class="c1">#      genes:                ## specify sets of genes to be colored</span>
<span class="c1">#        - red: nonMT        ## the first set of genes</span>
<span class="c1">#          green: Unspliced</span>
<span class="c1">#          blue: MT</span>
<span class="c1">#      # - ...               ## if more 1 set of genes are required</span>
<span class="c1">#      coord_per_pixel: 1000</span>
<span class="c1">#      auto_adjust: true</span>
<span class="c1">#      adjust_quantile: 0.99</span>
<span class="c1">#</span>
<span class="c1">#histology:                  ## specify the parameters for histology alignment using historef</span>
<span class="c1">#    min_buffer_size: 1000   ## min_buffer_size, max_buffer_size and step_buffer_size will create a list of buffer size help historef to do the alignment</span>
<span class="c1">#    max_buffer_size: 2000</span>
<span class="c1">#    step_buffer_size: 100</span>
<span class="c1">#    raster_channel: 1       ## roaster channel used for historef alignment</span>
<span class="c1">#</span>
<span class="c1">#downstream:</span>
<span class="c1">#  mu_scale: 1000           ## Specify coordinate to um translate for hexagon. By default, we consider the spatial digital gene expression matrix (SGE) is in nano meter.</span>
<span class="c1">#  gene_filter:             ## Specify the criteria for gene filtering in a manner compatible with regular expressions, which will be applied when creating the FICTURE-compatible SGE and hexagonal SGE.</span>
<span class="c1">#   keep_gene_type: &quot;protein_coding,lncRNA&quot;    # genes to keep</span>
<span class="c1">#   rm_gene_regex: &quot;^Gm\\d+|^mt-|^MT-&quot;         # genes to remove</span>
<span class="c1">#  segment:                 ## specify the parameters for grouping the pixels into hexagons</span>
<span class="c1">#    precision: 2           ## specify the number of digits to store spatial location (in um, 0 for integer)</span>
<span class="c1">#    min_pixel_per_unit: 10 ## specify a minimum UMI count of output hexagons</span>
<span class="c1">#    char:                  ## specify the characteristics for the hexagons</span>
<span class="c1">#      - solofeature: gn    ## specify the genomic feature to create hexagon</span>
<span class="c1">#        hexagonwidth: 24   ## specify the size for a hexagonal grid</span>
<span class="c1">#        segmentmove: 1     ## Specify if the SGE is based on overlapping hexagons or non-overlapping hexagon. The default value 1 will create non-overlapping hexagons.</span>
</code></pre></div></td></tr></table></div>
<h3 id="detailed-description-of-individual-fields">Detailed Description of Individual Fields<a class="headerlink" href="#detailed-description-of-individual-fields" title="Permanent link">&para;</a></h3>
<h4 id="input">Input<a class="headerlink" href="#input" title="Permanent link">&para;</a></h4>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p>NovaScope supports using relative paths in the job configuration file, which should be relative to the working directory. If a relative path is found, NovaScope automatically obtains its real path and uses it in the process.</p>
</div>
<ul>
<li>
<p><strong><code>seq1st</code></strong>:</p>
<ul>
<li>
<p><em><code>id</code></em>: The <code>id</code> will be used to organize the 1st-seq FASTQ files. Make sure the <code>id</code> parameter for 1st-seq in the corresponding flowcell is unique.  </p>
</li>
<li>
<p><em><code>layout</code></em>: A spatial barcode (sbcd) layout file to provide the layout of tiles in a chip with the following format. If absent, <a href="https://seqscope.github.io/NovaScope/">NovaScope</a> will automatically look for the sbcd layout within the NovaScope repository at <a href="https://github.com/seqscope/NovaScope/tree/main/info/assets/layout_per_tile_basis">info/assets/layout_per_tile_basis</a>, using the section chip ID for reference.
    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="l l-Scalar l-Scalar-Plain">lane  tile  row  col  rowshift  colshift</span>
<span class="l l-Scalar l-Scalar-Plain">3     2556  1    1    0         0</span>
<span class="l l-Scalar l-Scalar-Plain">3     2456  2    1    0         0.1715</span>
</code></pre></div></td></tr></table></div></p>
<ul>
<li><code>lane</code>: Lane IDs;</li>
<li><code>tile</code>: Tile IDs;</li>
<li><code>row</code> &amp; <code>col</code>: The layout position;</li>
<li><code>rowshift</code> &amp; <code>colshift</code>: The gap information</li>
</ul>
</li>
</ul>
</li>
<li>
<p><strong><code>seq2nd</code></strong>: This parameter requires all FASTQ pairs associated with the input section chip to be provided under <code>seq2nd</code>. </p>
<details class="note">
<summary>How to generate <code>seq2nd_pair_id</code>?</summary>
<p>If an ID is not specified, NovaScope will automatically generate one using the format <code>&lt;flowcell_id&gt;.&lt;chip_id&gt;.&lt;randomer&gt;</code>, where <code>randomer</code> is the last 5 digits of the md5 hash of the real path of the read 1 FASTQ file from the 2nd-seq.</p>
</details>
</li>
<li>
<p><strong><code>run_id</code></strong>: Only needed if alignment is required to generate the requested output. It is used as an identifier for alignment and Spatial Digital Gene Expression matrices (SGEs) to differentiate between input 2nd-seq FASTQ files. This is particularly useful when generating SGEs using the same 1st-seq files but different 2nd-seq files. If not provided, NovaScope will generate it based on the flowcell ID, chip ID, and all input 2nd-seq read 1 FASTQ files.</p>
<details class="note">
<summary>How to generate <code>run_id</code>?</summary>
<p>NovaScope automatically generates <code>run_id</code> in the format <code>&lt;flowcell_id&gt;-&lt;chip_id&gt;-&lt;species&gt;-&lt;randomer&gt;</code>. The <code>randomer</code> is created by sorting all input seq2nd_pair_id, concatenating these seq2nd_pair_id into a single long string, and then computing the md5 hash of this string. The last 5 digits of this hash are used as the <code>randomer</code>.</p>
</details>
</li>
<li>
<p><strong><code>unit_id</code></strong>: Only needed if reformat feature is required to generate the requested output. It acts as an identifier for SGEs that are prepared for reformatting. This identifier is especially useful when users wish to manually modify SGE outside of NovaScope and then proceed to reformat both the original and modified SGEs. The <code>unit_id</code> ensures clear distinction between the original and modified datasets.</p>
<details class="note">
<summary>How to generate <code>unit_id</code></summary>
<p>If <code>unit_id</code> is not specified and reformatting is requested, it will default to <code>&lt;run_id&gt;-default</code>, indicating that no manual preprocessing has occurred. </p>
<p>Users who prefer to reformat manually modified SGEs should define their own <code>unit_id</code>. We recommend incorporating <code>run_id</code> into the <code>unit_id</code> to maintain a clear trace of the dataset lineage.</p>
</details>
</li>
<li>
<p><strong><code>histology</code></strong>: NovaScope allows multiple input histology files for alignment. However, it is important to note that the magnification and type of each histology file serve as identifiers. Ensure that no two input histology files share the same magnification and type. Currently, <a href="https://github.com/seqscope/historef">historef</a> supports the following types:</p>
<ul>
<li><code>"hne"</code>: <a href="https://en.wikipedia.org/wiki/H%26E_stain">Hematoxylin and Eosin (H&amp;E) stained</a> histology images;</li>
<li><code>"dapi"</code>: <a href="https://en.wikipedia.org/wiki/DAPI">DAPI or 4',6-diamidino-2-phenylindole stained</a> histology images;</li>
<li><code>"fl"</code>: Fluorescence stained histology images.</li>
</ul>
</li>
</ul>
<h4 id="output">Output<a class="headerlink" href="#output" title="Permanent link">&para;</a></h4>
<p>The output directory will be used to organize the input files and store output files. Please see the structure directory <a href="../output/">here</a>.</p>
<h4 id="requests">Requests<a class="headerlink" href="#requests" title="Permanent link">&para;</a></h4>
<p>The pipeline interprets the requested output files via <code>request</code> and determines the execution flow.</p>
<div class="admonition info">
<p class="admonition-title">Info</p>
<p>The <code>request</code> parameter should indicate the <strong>final output</strong> required, and all intermediary files contributing to the final output will be automatically generated (i.e., the dependencies between rules). </p>
</div>
<p>Below are the options with their final output files and links to detailed output information. For more insights into the excution flow, please consult the <a href="../../walkthrough/execution_guide/rule_execution/">execution flow by request</a> alongside the <a href="https://seqscope.github.io/NovaScope/#an-overview-of-the-workflow-structure">rulegraph</a>. </p>
<table>
<thead>
<tr>
<th>Option</th>
<th>Main/Final Output Files</th>
<th>Details</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>sbcd-per-flowcell</code></td>
<td>Spatial barcode map (per-tile basis) and Manifest file for a flowcell</td>
<td><a href="../../walkthrough/rules/fastq2sbcd/#output-files">fastq2sbcd</a></td>
</tr>
<tr>
<td><code>sbcd-per-chip</code></td>
<td>Spatial barcode map for a section chip, Image of spatial barcode distribution</td>
<td><a href="../../walkthrough/rules/sbcd2chip/#output-files">sbcd2chip</a></td>
</tr>
<tr>
<td><code>smatch-per-chip</code></td>
<td>File with matched spatial barcodes, Image of matched barcode spatial distribution</td>
<td><a href="../../walkthrough/rules/smatch/#output-files">smatch</a></td>
</tr>
<tr>
<td><code>align-per-run</code></td>
<td>Binary Alignment Map (BAM) file, Digital gene expression matrix (DGE) for genomic features</td>
<td><a href="../../walkthrough/rules/align/">align</a></td>
</tr>
<tr>
<td><code>sge-per-run</code></td>
<td>Spatial digital gene expression matrix (SGE), Spatial distribution images for transcripts</td>
<td><a href="../../walkthrough/rules/dge2sdge/">dge2sdge</a> and <a href="../../walkthrough/rules/sdge_visual/">sdge_visual</a></td>
</tr>
<tr>
<td><code>hist-per-run</code></td>
<td>Geotiff files for coordinate transformation between SGE and histology image, and a resized one</td>
<td><a href="../../walkthrough/rules/historef/">historef</a></td>
</tr>
<tr>
<td><code>transcript-per-unit</code></td>
<td>SGE in in the FICTURE-compatible format</td>
<td><a href="../../walkthrough/rules/sdgeAR_reformat/">sdgeAR_reformat</a></td>
</tr>
<tr>
<td><code>segment-per-unit</code></td>
<td>Hexagon-based SGE in the 10x genomics format</td>
<td><a href="../../walkthrough/rules/sdgeAR_segment/">sdgeAR_segment</a></td>
</tr>
</tbody>
</table>
<h4 id="upstream-downstream">Upstream &amp; Downstream<a class="headerlink" href="#upstream-downstream" title="Permanent link">&para;</a></h4>
<p>Parameter details for the <code>upstream</code> and <code>downstream</code> fields are outlined in the <a href="../../walkthrough/intro/">NovaScope Walkthrough</a>, under the specific rule pages to which they apply.</p>
<ul>
<li><strong><code>align</code></strong><ul>
<li>
<p><em><code>resource</code></em>: Only applicable for HPC users.</p>
<ul>
<li><code>assign_type</code>: two available options for how NovaScope allocates resources for alignment. The options include <code>"stdin"</code> (recommended) and <code>"filesize"</code>. Details for each option are provided in the blocks below. </li>
</ul>
<details class="note">
<summary>Option <code>stdin</code></summary>
<p><strong>Advantages:</strong>
- Directly allocates resources as specified in the <code>stdin</code> field, bypassing calculations for precision in resource management.
- Enables customization of resources for different datasets in the job configuration file, allowing for optimization of costs based on file size.</p>
<p><strong>Disadvantages:</strong>
- Requires users to specify resources for each job unless default settings (partition name, threads, memory) fit the computing environment. An example is provided in the <a href="#a-template-of-the-config-file">template</a>.</p>
</details>
<details class="note">
<summary>Option <code>filesize</code></summary>
<p><strong>Advantages:</strong>
- Automatically allocates resources based on the total size of input 2nd-seq FASTQ files and specified computing resources in the <a href="../../installation/env_setup/#optional-computing-capabilities">environment configuration file</a>.
- Once computing resources are specified in the environment file, they automatically apply to all jobs, simplifying the setup.</p>
<p><strong>Disadvantages:</strong>
- Requires computing time to calculate the total size of input files, potentially delaying the start of data processing.</p>
<p>The resource allocation strategy is as follows:</p>
<table>
<thead>
<tr>
<th>Total File Size (GB)</th>
<th>Memory Allocated for Alignment (GB)</th>
</tr>
</thead>
<tbody>
<tr>
<td>Under 200</td>
<td>70</td>
</tr>
<tr>
<td>200 to 400</td>
<td>140</td>
</tr>
<tr>
<td>Over 400</td>
<td>330</td>
</tr>
</tbody>
</table>
</details>
</li>
</ul>
</li>
</ul>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.indexes", "navigation.sections", "content.code.annotate"], "search": "../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.c8d2eff1.min.js"></script>
      
        <script src="../../js/open_in_new_tab.js"></script>
      
    
  </body>
</html>